joblib/test/test_parallel.py

```diff
@@ -940,3 +940,32 @@ def __reduce__(self):
     with warns(DeprecationWarning):
         with raises(ValueError):
             delayed(UnpicklableCallable(), check_pickle=True)
+
+
+def _backend_reset_func(x):
```

Comment:
A slightly simpler snippet I ran locally a few times to make sure it was failing as expected (on master).

```py
import time

from joblib import Parallel, delayed


backend = 'multiprocessing'
n_jobs = 2
n_inputs = 1000
task_time = 1. / n_inputs

p = Parallel(n_jobs=n_jobs, backend=backend)

t0 = time.time()
Parallel(n_jobs=n_jobs, backend=backend)(
    delayed(time.sleep)(task_time) for i in range(n_inputs))
print(time.time() - t0)

t0 = time.time()
p(delayed(time.sleep)(task_time) for i in range(n_inputs))
print(time.time() - t0)

t0 = time.time()
p(delayed(time.sleep)(task_time) for i in range(n_inputs))
print(time.time() - t0)
```

I think you should compare first run and second run, make sure they are close (with some hard-coded relative tolerance) and same thing for the second and third run.